static/app/components/notificationActions/notificationActionManager.tsx

```diff
@@ -0,0 +1,203 @@
+   */
+  actions: NotificationAction[];
+  /**
+   * The list of available notification actions
+   */
+  availableActions: AvailableNotificationAction[];
+  /**
+   * The project associated with the notification actions
+   * TODO(enterprise): refactor to account for multiple projects
+   */
+  project: Project;
+  /**
+   * Updates the notification alert count for this project
+   */
+  updateAlertCount: (projectId: number, alertCount: number) => void;
+  /**
+   * Optional list of roles to display as recipients of Sentry notifications
+   */
+  recipientRoles?: string[];
+};
+
+const NotificationActionManager = ({
+  actions,
+  availableActions,
+  recipientRoles,
+  project,
+  updateAlertCount,
+}: NotificationActionManagerProps) => {
+  const [notificationActions, setNotificationActions] =
+    useState<Partial<NotificationAction>[]>(actions);
+
+  const addNotificationAction = (
+    actionInstance: AvailableNotificationAction['action']
+  ) => {
+    const updatedActions = [...notificationActions, actionInstance];
+    setNotificationActions(updatedActions);
+    updateAlertCount(parseInt(project.id, 10), updatedActions.length);
+  };
+
+  const removeNotificationAction = (index: number) => {
+    // Removes notif action from state using the index
+    const updatedActions = [...notificationActions];
+    updatedActions.splice(index, 1);
+    setNotificationActions(updatedActions);
+    updateAlertCount(parseInt(project.id, 10), updatedActions.length);
+  };
+
+  const updateNotificationAction = (index: number, updatedAction: NotificationAction) => {
+    // Updates notif action from state using the index
+    const updatedActions = [...notificationActions];
+    updatedActions.splice(index, 1, updatedAction);
+    setNotificationActions(updatedActions);
+  };
+
+  // Lists the available actions for each service
+  const availableServices: Record<
+    NotificationActionService,
+    AvailableNotificationAction[]
+  > = {
+    [NotificationActionService.SENTRY_NOTIFICATION]: [],
+    [NotificationActionService.EMAIL]: [],
+    [NotificationActionService.SLACK]: [],
+    [NotificationActionService.PAGERDUTY]: [],
+    [NotificationActionService.MSTEAMS]: [],
+    [NotificationActionService.SENTRY_APP]: [],
+  };
+  availableActions.forEach(a => {
+    availableServices[a.action.serviceType as NotificationActionService].push(a);
+  });
+
+  // Groups the notification actions together by service
+  // Will render the notif actions in the order the keys are listed in
+  const actionsMap: Record<
+    NotificationActionService,
+    {action: NotificationAction; index: number}[]
+  > = {
+    [NotificationActionService.SENTRY_NOTIFICATION]: [],
+    [NotificationActionService.EMAIL]: [],
+    [NotificationActionService.SLACK]: [],
+    [NotificationActionService.PAGERDUTY]: [],
+    [NotificationActionService.MSTEAMS]: [],
+    [NotificationActionService.SENTRY_APP]: [],
+  };
+  notificationActions.forEach((action, index) => {
```

Comment:
same comment - memoize probably or use state

Reply:
@nhsiehgit Do we have to memo everything? This comment on my last PR has some fair points: https://github.com/getsentry/sentry/pull/46613#discussion_r1153931357, especially the [linked article ](https://maxrozen.com/understanding-when-use-usememo#when-to-use-usememo).

I'm not idealogical about memo or no-memo, but IMO simpler code is generally better.

Reply:
mmmm that's fine i guess
definitely a premature optimization but imo depending on the use case, still makes complicated components a LOT easier to understand the lifecycle.

IMO `useMemo` is a cheap and easy way to encapsulate heavy logic and introduce a little organization/code cleanliness which of course has subjective value. 
So hence i'm 🤷 on this. It's not something i'll block on, but I disagree that it's 0 value add
To the articles point - this is a tradeoff btw memory and cpu.

Reply:
I don't know if i commented anywhere, but anything where we're looping through things that we _don't_ need to recalculate every render, i don't see any reason why _not_ to memoize